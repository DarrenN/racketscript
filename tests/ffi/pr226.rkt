#lang racketscript/base

(require racketscript/interop
         racket/string
         (for-syntax racket/base syntax/parse racket/syntax))

(define-syntax define-js-checker
  (syntax-parser
    [(_ name #:checks-js thing)
     #:with thing-pred? (format-id #'thing "js-~a?" #'thing)
     #'(define-syntax name
         (syntax-parser
           [(_ e) #'(name e "")]
           [(_ e label)
            #'(let ([x e])
                (when (non-empty-string? label)
                  (printf "~a " label))
                (printf "~v is a js ~a?: ~a\n" x 'thing (thing-pred? x)))]))]))

(define-js-checker check-js-obj #:checks-js object)
(define-js-checker check-js-array #:checks-js array)

(define-syntax check-js-objs+arrays
  (syntax-parser
    [(_) #'(begin)]
    [(_ e #:label str . rst)
     #'(begin
         (let ([x e])
           (check-js-obj x str)
           (check-js-array x str))
         (check-js-objs+arrays . rst))]
    [(_ e . rst) ; no label
     #'(begin
         (let ([x e])
           (check-js-obj x)
           (check-js-array x))
         (check-js-objs+arrays . rst))]))

(check-js-objs+arrays {$/obj}
                      [$/array 1 2 3] #:label "raw js array"
                      (make-hash)
                      (vector 1 2 3)
                      #"ABC" #:label "racket bytestring" ; Uint8Array TypedArray is not Array
                      #\D #:label "racket char"
                      "a racket string"
                      #js"a js string"
                      '()
                      '(1 2 3)
                      1
                      #f
                      $/null)

