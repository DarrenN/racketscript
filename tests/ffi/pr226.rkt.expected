[object Object] is a js object?: #t
[object Object] is a js array?: #f
raw js array 1,2,3 is a js object?: #t
raw js array 1,2,3 is a js array?: #t
'#hash() is a js object?: #f
'#hash() is a js array?: #f
'#(1 2 3) is a js object?: #f
'#(1 2 3) is a js array?: #f
racket bytestring #"ABC" is a js object?: #t
racket bytestring #"ABC" is a js array?: #f
racket char #\D is a js object?: #f
racket char #\D is a js array?: #f
"a racket string" is a js object?: #f
"a racket string" is a js array?: #f
a js string is a js object?: #f
a js string is a js array?: #f
'() is a js object?: #f
'() is a js array?: #f
'(1 2 3) is a js object?: #f
'(1 2 3) is a js array?: #f
1 is a js object?: #f
1 is a js array?: #f
#f is a js object?: #f
#f is a js array?: #f
#<void> is a js object?: #f
#<void> is a js array?: #f
